modules/core/src/main/java/org/apache/ignite/internal/processors/diagnostic/ReconciliationExecutionContext.java

```diff
@@ -0,0 +1,152 @@
+import org.apache.ignite.internal.managers.communication.GridIoPolicy;
+
+/**
+ * Responsible for storing context of all ongoing reconciliation activities.
+ */
+public class ReconciliationExecutionContext {
+    /** Session ID for test purposes, permits won't be checked if it's passed. */
+    public static final long IGNORE_JOB_PERMITS_SESSION_ID = Long.MIN_VALUE / 7;
+
+    /** Maximum number of stored sessions. */
+    private static final int MAX_SESSIONS = 10;
+
+    /**Kernal context. We need only closure processor from it, but it's not set at the moment of initialization. */
+    private final GridKernalContext kernalCtx;
+
+    /** Node local running jobs limit according to the parallelism level. */
+    private final Map<Long/*Session ID*/, Integer> runningJobsLimit = new LinkedHashMap<>();
+
+    /** Number of currently running reconciliation jobs on the local node. */
+    private final Map<Long/*Session ID*/, Integer> runningJobsCnt = new LinkedHashMap<>();
+
+    /** Jobs that were enqueued to be executed later due to permits exhaustion. */
+    private final Map<Long/*Session ID*/, Queue<ComputeJobContinuation>> pendingJobs = new LinkedHashMap<>();
+
+    /** Id of last or current reconciliation session. */
+    private long sesId;
+
+    /**
+     * @param kernalCtx Kernal context.
+     */
+    public ReconciliationExecutionContext(GridKernalContext kernalCtx) {
+        this.kernalCtx = kernalCtx;
+    }
+
+    /**
+     * @return Id of last or current reconciliation session.
+     */
+    public synchronized long sessionId() {
+        return sesId;
+    }
+
+    /**
+     * Registers new partitions reconciliation session.
+     *
+     * @param sesId Session ID.
+     * @param parallelism Parallelism level.
+     */
+    public synchronized void registerSession(long sesId, int parallelism) {
+        this.sesId = sesId;
+
+        runningJobsCnt.put(sesId, 0);
+
+        runningJobsLimit.put(sesId, parallelism);
+
+        pendingJobs.put(sesId, new LinkedList<>());
+
+        if (runningJobsCnt.size() == MAX_SESSIONS + 1) {
+            runningJobsCnt.entrySet().iterator().remove();
+
+            runningJobsLimit.entrySet().iterator().remove();
+
+            pendingJobs.entrySet().iterator().remove();
+        }
+
+
+    }
+
+    /**
+     * Acquires permit for the job execution or holds its execution if permit is not available.
+     *
+     * @param sesId Session ID.
+     * @param jobCont   Job context.
+     * @return <code>true</code> if the permit has been granted or
+     * <code>false</code> if the job execution should be suspended.
+     */
+    public synchronized boolean acquireJobPermitOrHold(long sesId, ComputeJobContinuation jobCont) {
+        if (sesId == IGNORE_JOB_PERMITS_SESSION_ID)
+            return true;
+
+        int limit = runningJobsLimit.get(sesId);
+
+        int running = runningJobsCnt.get(sesId);
+
+        if (running < limit) {
+            runningJobsCnt.put(sesId, running + 1);
```

Comment:
compute?)

Reply:
put is fine too in synchronized context.